Improve performance and fix limitations of flatten_byte_operators [blocks: #3725]

tautschnig opened the pull request:
~Depends on #2061 and #2063 and thus should not be merged just yet.~

smowton commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,49 @@ bvt boolbvt::convert_member(const member_exprt &expr)
+#endif
+
+        error().source_location = expr.find_source_location();
+        error() << "member: component type does not match: " << subtype.pretty()
```
Recommend newlines before `pretty` for readability

smowton commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,49 @@ bvt boolbvt::convert_member(const member_exprt &expr)
+        error().source_location = expr.find_source_location();
+        error() << "member: component type does not match: " << subtype.pretty()
+                << " vs. " << expr.type().pretty() << eom;
+        throw 0;
```
I think this is an invariant, as it's our responsibility to make sure this is true, not a user error

smowton commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,49 @@ bvt boolbvt::convert_member(const member_exprt &expr)
-
-        return bv;
+#if 0
+        std::cout << "DEBUG " << expr.pretty() << "\n";
```
If we're going to commit this, replace DEBUG with a more useful key?

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+  mp_integer num_elements;
+  if(to_integer(max_bytes, num_elements) && to_integer(src_size, num_elements))
+  {
+    throw "cannot unpack array/vector of non-const size:\n" +
```
Again, invariant?

tautschnig replied:
It's something that may be fixable in some way, and really a limitation of the implementation rather than a bug. So I'm inclined to leave it in place.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+
+  mp_integer offset_int = 0;
+
+  if(element_width > 0 && element_width % 8 == 0)
```
What would `element_width == 0` mean?

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+  array_exprt &dest_array)
+{
+  mp_integer num_elements;
+  if(to_integer(max_bytes, num_elements) && to_integer(src_size, num_elements))
```
!!! These two appear to have different units -- bytes and array elements respectively?

tautschnig replied:
Fixed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+    mp_integer el_bytes = element_width / 8;
+
+    if(num_elements % el_bytes != 0)
+      num_elements += el_bytes;
```
This appears to be a number of bytes at this point, but was taken as a number of elements above?

tautschnig replied:
Fixed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+    if(num_elements % el_bytes != 0)
+      num_elements += el_bytes;
+
+    num_elements /= el_bytes;
```
Now it's a number of elements. Please don't use something named `_elements` to store a byte count

tautschnig replied:
Fixed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+    if(!to_integer(offset, offset_int))
+    {
+      // compute offset as number of elements
+      offset_int /= el_bytes;
```
`offset_elements`?

tautschnig replied:
Fixed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+
+  exprt src_simp = simplify_expr(src, ns);
+
+  for(mp_integer i = offset_int; i < num_elements; ++i)
```
Given the above units trouble, `i` -> `element_idx`?

tautschnig replied:
I think I've made a right mess here. Thanks for looking carefully, I'm working on cleanup.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -17,20 +19,106 @@ Author: Daniel Kroening, kroening@kroening.com
+
+    if(src_simp.id() != ID_array || i >= src_simp.operands().size())
+    {
+      index = index_exprt(src_simp, from_integer(i, index_type()));
```
`index` -> `element`?

tautschnig replied:
Done.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -100,13 +185,109 @@ static exprt unpack_rec(
+          new_offset = from_integer(new_offset_int, offset.type());
+      }
+
+      if(!to_integer(max_bytes, new_max_bytes_int))
```
Comment before these what the intended adjustment is? At the moment it looks like you would reduce *both* lower and upper bounds if the access is not byte-aligned, whereas I might expect you to round the lower bound down and the upper bound up to ensure you include the desired range?

tautschnig replied:
No, byte-alignedness is tested above, there shouldn't be a need for rounding I believe.

smowton replied:
In that case `member_offset_bits % 8` is always zero, as it starts at zero and is the incremental sum of several `element_width`s, each of which is itself a multiple of eight.

tautschnig replied:
Your comment really highlighted that there was a problem: `member_offset_bits` wasn't actually being used properly, this should have been a division, not modulo!

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -227,29 +414,63 @@ exprt flatten_byte_extract(
+      UNREACHABLE;
+
+    // TODO: consider ways of dealing with vectors of unknown subtype
+    // size or with a subtype size that does not fit byte boundaries
```
What currently happens?

smowton reviewed (commented):
Reviewed up to `byte_operators.cpp:578`, to continue tomorrow

romainbrenguier commented on src/solvers/flattening/boolbv_byte_update.cpp:
```diff
@@ -8,18 +8,31 @@ Author: Daniel Kroening, kroening@kroening.com
+    has_subtype(expr.op0().type(), ID_pointer, ns))
+  {
+    exprt tmp = lower_byte_update(expr, ns);
+    return convert_bv(tmp);
```
`const exprt tmp` or just one line return

romainbrenguier commented on src/solvers/flattening/boolbv_equality.cpp:
```diff
@@ -30,7 +30,7 @@ literalt boolbvt::convert_equality(const equal_exprt &expr)
     if(has_byte_operator(expr))
     {
-      exprt tmp=flatten_byte_operators(expr, ns);
+      exprt tmp = lower_byte_operators(expr, ns);
```
const

romainbrenguier commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,41 @@ bvt boolbvt::convert_member(const member_exprt &expr)
-        it++)
+  for(struct_union_typet::componentst::const_iterator it = components.begin();
+      it != components.end();
+      ++it)
```
Use ranged for if `it` is not used as an iterator inside the loop

romainbrenguier commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,41 @@ bvt boolbvt::convert_member(const member_exprt &expr)
-      offset+=sub_width;
+      DATA_INVARIANT(
+        base_type_eq(subtype, expr.type(), ns),
+        "component type expected to match expression type:\n" + expr.pretty());
```
I think it's better to avoid calling `pretty` in places where it will be called even if there is no error.

tautschnig replied:
I'm not sure I understand the "called even if there is no error" part - this would only be printed in case of a violated invariant?!

romainbrenguier replied:
oh yes sorry, you can ignore my comment, the macro is so that the second part does not need to be called.

romainbrenguier commented on src/solvers/flattening/boolbv_member.cpp:
```diff
@@ -20,67 +21,41 @@ bvt boolbvt::convert_member(const member_exprt &expr)
+        offset + sub_width <= struct_bv.size(), "member access outside struct");
+
+      for(std::size_t i = 0; i < sub_width; i++)
+        bv[i] = struct_bv[offset + i];
```
I think you could use `bv.insert(bv.begin(), struct_bv + offset, struct_bv + offset + sub_width)` instead

tautschnig replied:
Good call! This actually enables a further simplification of just doing `return bvt(struct_bv.begin() + ...)`.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -323,8 +590,22 @@ exprt flatten_byte_extract(
+
+    mp_integer index = 0;
+    if(
+      !offset_i.is_constant() || root.id() != ID_array ||
```
Readability: suggest break after each disjunct

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -371,41 +657,40 @@ exprt flatten_byte_update(
       if(sub_size==1)
       {
+        byte_extract_exprt byte_extract_expr(
+          src.id()==ID_byte_update_little_endian?
```
might as well use conventional operator spacing here

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -531,23 +816,24 @@ exprt flatten_byte_update(
-          src.op2(), t);
-    else
-      value_extended=src.op2();
+          value_extended, unsignedbv_typet(width));
```
Style: params all on one line, or one per line

smowton reviewed (approved):
Thanks for fixing the comments above; done reviewing with only nitpicks regarding the second half.

smowton commented on regression/cbmc/memcpy4/main.c:
```diff
@@ -0,0 +1,16 @@
+  char foo[1];
+  char result = 0;
+  int size;
+  __CPROVER_assume(size == 1);
```
Worth commenting that this weird construct is here to defeat constant propagation (I presume?) and shouldn't be optimised away.

smowton commented on regression/cbmc/union10/main.c:
```diff
@@ -0,0 +1,23 @@
+  int c;
+} S2;
+
+static S1 (* const l[1]) = {((S1 *) ((0x50000000UL) + 0x00000) )};
```
`0x00000`? Just `0`? Also why, won't cbmc just optimise `0x50000000 + 0` to eliminate the pointless +0?

smowton replied:
Maybe it's just me, but this declaration baffled me -- the bracketing around `* const l[1]` suggested a function? Don't know if there's a more obvious way to write the type "array of const S1*s" though.

smowton commented on regression/cbmc/union10/main.c:
```diff
@@ -0,0 +1,23 @@
+void main()
+{
+  __CPROVER_allocated_memory(0x50000000UL, sizeof(S1));
+  l[0]->access[0] = 0;
```
Why is l a single-item array instead of a simple pointer? Are you trying to defeat optimisation or to exercise some particular behaviour? Either way add a comment clarifying what you're trying to do.

smowton commented on regression/cbmc/union10/main.c:
```diff
@@ -0,0 +1,23 @@
+  __CPROVER_assert(l[0]->access[0] == 0, "holds");
+  __CPROVER_assert(l[0]->access[1] == 0, "should fail");
+  __CPROVER_allocated_memory(0x40000000UL + 0x40000, sizeof(S2));
+  ((S2 *) ((0x40000000UL) + 0x40000))->c = 0x0707;
```
Intending to trip an out-of-bounds write error? If so comment.

smowton reviewed (commented):
Some comment requests for the tests, haven't reviewed the code.

tautschnig committed:
Lower byte_{extract,update} over pointers and unbounded arrays

The array theory does not handle byte_{extract,update} operators. For pointers,
make sure the flattening back-end also uses lowering rather than failing with an
exception.

tautschnig committed:
Do not introduce byte_extract_exprt during flattening

A prior lowering pass may have removed them (for example, if they occur within
unbounded arrays).

tautschnig committed:
Rewrite byte_update lowering to respect endianness, support composite types

byte_update lowering now proceeds as follows:
1) Determine the size of the update, with the size of the object to be
updated as an upper bound. We fail if neither can be determined.
2) Turn the update value into a byte array of the size determined above.
3) If the offset is not constant, turn the object into a byte array, and
use a "with" expression to encode the update; else update the values in
place.
4) Construct a new object.

tautschnig commented:
This (hopefully!) final part of the byte-operator rework is now ready for review. I didn't see a way to step-by-step fix the byte_update lowering code and instead rewrote it completely. I'm afraid this doesn't make for a particularly pleasant review of the first commit in here. Consequently it may be more helpful to subject this branch to extensive testing. It is perfectly possible that with some of the PRs depending on this one we still run into issues, but then at least the implementation should now be amenable to local fixes, depending on what input data types we are looking at.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 96d9f792b698ae0da504b7b5a72144b2402b44b3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101248909

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b05c96c070d40ac89556badee4f583c49c1adf58).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101272202

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+  exprt result = src.op();
+
+  std::size_t i = 0;
+  for(const auto &element : value_as_byte_array.operands())
```
Don't use ranged-for since you need the index anyway

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    }
+    else
+    {
+      if_exprt conditional_element(
```
Brace style initialization  is preferred.

tautschnig replied:
Done - I hope I have caught all of them, but it's very well possible that I have missed one.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+        continue;
+    }
+
+    if(!non_const_update_bound.has_value())
```
Note: it's better to avoid late decision making in general

tautschnig replied:
I have reorganised this code a bit, but I'm not sure whether this is the improvement you are looking for. (I have introduced an `exprt update_value`.)

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+      *non_const_update_bound, subtype_size.type());
+  else
+    update_bound =
+      from_integer(value_as_byte_array.operands().size(), initial_bytes.type());
```
use braces when the instruction in the `if ... else ...` spread over several lines

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+
+  // encode the first update: update the original element at first_index with
+  // initial_bytes-many bytes extracted from value_as_byte_array
+  with_exprt result(
```
brace style initialization preferred

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+  const optionalt<exprt> &non_const_update_bound,
+  const namespacet &ns);
+
+static exprt lower_byte_update_byte_array_vector(
```
New function should be documented

tautschnig replied:
Done (all newly added functions are now documented).

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+}
 
-  const mp_integer &element_size = *element_size_opt;
+static exprt lower_byte_update_array_vector_non_const(
```
New function should be documented

tautschnig replied:
Done (all newly added functions are now documented).

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+  return simplify_expr(result, ns);
+}
+
+static exprt lower_byte_update_array_vector(
```
new function should be documented

tautschnig replied:
Done (all newly added functions are now documented).

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    if(
+      (offset_bytes + value_as_byte_array.operands().size()) * 8 <=
+        i * *subtype_bits ||
+      offset_bytes * 8 >= (i + 1) * *subtype_bits)
```
Part of the computation could be taken out of the loop and it looks like the loop should be split to avoid unnecessary checks: once the first condition is true then it will always be true?

tautschnig replied:
Indeed, I have split the loop into three chunks.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    mp_integer update_elements = *subtype_bits / 8;
+    exprt::operandst::const_iterator first =
+      value_as_byte_array.operands().begin();
+    exprt::operandst::const_iterator end = value_as_byte_array.operands().end();
```
These could be taken out of the loop

tautschnig replied:
No, they are modified below depending on the value of `update_offset`, which in turn depends on the loop counter.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
-      const mp_integer &sub_size = *sub_size_opt;
+    if(std::distance(first, end) > update_elements)
+      end = first + numeric_cast_v<std::ptrdiff_t>(update_elements);
+    exprt::operandst update_values(first, end);
```
brace initialization preferred

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    exprt::operandst update_values(first, end);
+    const std::size_t update_size = update_values.size();
+
+    byte_update_exprt bu(
```
brace initialization preferred

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    exprt::operandst update_values(first, end);
+    const std::size_t update_size = update_values.size();
+
+    byte_update_exprt bu(
```
const?

tautschnig replied:
Done.

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+      vector_exprt(std::move(elements), to_vector_type(src.type())), ns);
+}
+
+static exprt lower_byte_update_struct(
```
new function should be documented

tautschnig replied:
Done (all newly added functions are now documented).

romainbrenguier commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+
+    // compute the update offset relative to this struct member - will be
+    // negative if we are already in the middle of the update or beyond it
+    exprt offset = simplify_expr(
```
const?

tautschnig replied:
No, it is conditionally reset to 0 later on.

romainbrenguier commented on src/solvers/flattening/boolbv_byte_update.cpp:
```diff
@@ -8,16 +8,39 @@ Author: Daniel Kroening, kroening@kroening.com
+#include "bv_conversion_exceptions.h"
 #include "bv_endianness_map.h"
 
 bvt boolbvt::convert_byte_update(const byte_update_exprt &expr)
```
the fact that the method can throw an exception should probably be documented

romainbrenguier commented on src/solvers/lowering/expr_lowering.h:
```diff
@@ -12,11 +12,14 @@ Author: Michael Tautschnig
 
 exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns);
 
+exprt lower_byte_update(const byte_update_exprt &src, const namespacet &ns);
```
Ideally should be documented. Also I don't find the commit message very clear (what's the change being made?).

tautschnig replied:
Done (all newly added functions are now documented). Rewrote the commit message as well.

tautschnig committed:
Use byte_update lowering in SMT2 conversion

The locally implemented lowering was very incomplete.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 45c612f8c5e427e3efff4d9d0b678ac16b5fd7f1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101349527

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+        element,
+        index_exprt(src.op(), where));
+
+      if(result.id() != ID_with)
```
Could factor this better to be like
```
get element
if(maybe out of range)
  add if-guard
if(already with)
  add to with
else
  make new with
```

tautschnig replied:
Done as proposed.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+static exprt lower_byte_update_array_vector_non_const(
+  const byte_update_exprt &src,
+  const typet &subtype,
+  const array_exprt &value_as_byte_array,
```
PRECONDITION that this really is made of chars?

tautschnig replied:
Do you feel this would add much value? I can of course do it, but I think things would either not break or break in an obvious way if that ever were violated.

smowton replied:
Sure, no big deal

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+  // update values that will be consumed in each step; if we cannot determine a
+  // constant value at this time we assume it's just one byte
+  std::size_t step_size = 1;
+  if(subtype_size.is_constant())
```
I'm confused what you intend to happen / how this is supposed to work when this is non-constant; suggest commenting to that effect.

tautschnig replied:
Code amended to use a symbolic value and comment clarified.

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+          extract_opcode,
+          value_as_byte_array,
+          from_integer(offset, src.offset().type()),
+          array_typet(unsignedbv_typet(8), subtype_size))),
```
It's a little surprising to see `byte_extract` used to pick up a slice of a byte-array; is `extractbits_exprt` better for this, or just an `array_exprt` of the relevant bytes? It's not better or worse particularly, I just note we don't need byte-extract's casting powers here.

tautschnig replied:
I need/want the power of a possibly non-constant size of the slice, hence falling back to `byte_extract` (which in turn is subjected to lowering, which might result in either an `extractbits_exprt` or `array_exprt`).

smowton commented on src/solvers/lowering/byte_operators.cpp:
```diff
@@ -619,267 +622,638 @@ exprt lower_byte_extract(const byte_extract_exprt &src, const namespacet &ns)
+    index_exprt index(src.op(), from_integer(i, index_type()));
+
+    if(
+      (offset_bytes + value_as_byte_array.operands().size()) * 8 <=
```
Suggest a comment giving the intended meaning (if element fully in range?)

tautschnig replied:
Comments added while splitting up the loop into chunks.

smowton reviewed (commented):
Some comments for what it's worth, but I'm not particularly confident of having accurately reviewed this because there's a lot of very similar edge-case code here. I'd suggest torturing the code with nasty edge-case byte-update operations (as unit tests I guess, since it's easier to be sure you're giving it the form of expression you expect), rather than relying on manual review here. If you're convinced you've already got good unit test coverage then lgtm.

tautschnig committed:
Byte-operator lowering: support void-typed updates

These may be generated when dereferencing invalid pointers, and are just no-ops.
Make sure we don't fail as size_of_expr would return nil_exprt() for those.

tautschnig commented:
@romainbrenguier @smowton I believe I have incorporated all feedback.

tautschnig committed:
Byte-operator lowering: support structs containing bit-fields

This removes the constraint on aligned member accesses. Includes factoring out
of unpack_struct to avoid growing the size of unpack_rec even further.

smowton commented:
Restarted spurious CI failure -- is this good to go with @kroening or @peterschrammel's review?

tautschnig commented:
> is this good to go with @kroening or @peterschrammel's review?

Yes, reviews would be much appreciated.

kroening commented on src/solvers/flattening/boolbv_byte_extract.cpp:
```diff
@@ -40,8 +41,12 @@ bvt map_bv(const bv_endianness_mapt &map, const bvt &src)
+    is_unbounded_array(expr.op().type()) ||
+    has_subtype(expr.type(), ID_pointer, ns) ||
+    has_subtype(expr.op().type(), ID_pointer, ns))
   {
```
What's the rationale for that, actually?
Is the alternative to always use the flattening code?

tautschnig replied:
The rationale for lowering byte-operators when pointers are involved? Given that there is another commit that does lowering when encountering a pointer I can check whether that's still necessary (without the latter it would certainly lead to an exception)~, it's a matter of whether the SMT back-end now needs this or not~(edit: this code is not for SMT).

tautschnig replied:
Very good call, this is now unnecessary. Commit (for both byte extract and byte update) and commit message amended.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4ada4bf09e6ad6dcb52ba6befd36d82ca4bf1093).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102380441

kroening commented:
Let's merge this now -- in the long run, it may be the case that the bit-vector flattener in SMT2 possibly outperforms the lowerer for byte_update.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 34a4de990725080f78007368d586026c69349109).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102381986
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
